relayer/relayer/src/trader/mod.rs

```diff
@@ -0,0 +1,302 @@
+    MissingRequired(String),
+
+    #[error("Address is not an AccountId {0}")]
+    NotAccountId(String),
+
+    #[error("Trader has encountered an unexpected error")]
+    Unexpected,
+}
+
+#[derive(Copy, Clone)]
+pub struct Trader;
+
+impl Trader {
+    pub async fn run(
+        config: Arc<Config>,
+        azero_signed_connection: Arc<AzeroConnectionWithSigner>,
+        eth_signed_connection: Arc<SignedEthConnection>,
+        mut circuit_breaker_receiver: broadcast::Receiver<CircuitBreakerEvent>,
+    ) -> Result<CircuitBreakerEvent, TraderError> {
+        let Config {
+            azero_contract_metadata,
+            azero_contract_address,
+            azero_ref_time_limit,
+            azero_proof_size_limit,
+            router_address,
+            router_metadata,
+            azero_wrapped_azero_address,
+            azero_ether_address,
+            azero_ether_metadata,
+            eth_wrapped_ether_address,
+            ..
+        } = &*config;
+
+        let most_azero = MostInstance::new(
+            azero_contract_address,
+            azero_contract_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let router = RouterInstance::new(
+            &router_address
+                .clone()
+                .ok_or(TraderError::MissingRequired("router_address".to_owned()))?,
+            router_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let azero_ether_address =
+            azero_ether_address
+                .clone()
+                .ok_or(TraderError::MissingRequired(
+                    "azero_ether_address".to_owned(),
+                ))?;
+
+        let azero_ether = AzeroEtherInstance::new(
+            &azero_ether_address,
+            azero_ether_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let wrapped_azero_address =
+            AccountId::from_str(&azero_wrapped_azero_address.clone().ok_or(
+                TraderError::MissingRequired("azero_wrapped_azero_address".to_owned()),
+            )?)
+            .map_err(|err| TraderError::NotAccountId(err.to_owned()))?;
+
+        let wrapped_ether_address = eth_wrapped_ether_address
+            .clone()
+            .ok_or(TraderError::MissingRequired(
+                "eth_wrapped_ether_address".to_owned(),
+            ))?
+            .parse::<Address>()?;
+
+        let wrapped_ether = WETH9::new(wrapped_ether_address, eth_signed_connection.clone());
+
+        let whoami_azero = azero_signed_connection.account_id();
+        let whoami_eth = eth_signed_connection.address().to_string();
+
+        info!("Starting");
+
+        select! {
+            cb_event = circuit_breaker_receiver.recv() => {
+                warn!("Exiting due to a circuit breaker event {cb_event:?}.");
+                Ok(cb_event?)
+            },
+
+            _ = async {
```

Comment:
Please use the result - in the current version all info about the error that occurred in Trader is lost and we get a mysterious `Unexpected`